Keyboard quality of life

kov opened the pull request:
A couple of changes to improve quality of life when using keyboard shortcuts on non-Mac VMs. With these you will be able to let, say, Gnome handle both Command-Tab and Control-arrows instead of them getting stolen by Mac OS. It also swaps Option and Command by default, so that you don't need to adjust your muscle memory to the new positions of Super and Alt (and don't need to swap them at the Mac OS level either).

I am submitting this set to your branch as it is based on top of your work that you are still upstreaming. Maybe you can submit them along with the rest of your work, otherwise I will wait for 'Divide ui/cocoa' patch to get upstreamed so that I can submit these two.

akihikodaki committed:
ui/cocoa: Set UI information

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
ui/cocoa: Use NSWindow's ability to resize

This change brings two new features:
- The window will be resizable if "Zoom To Fit" is eanbled
- The window can be made full screen by clicking full screen button
  provided by the platform. (The left-top green button.)

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
ui/cocoa: Do not scale UI without "Zoom To Fit"

Cocoa scales UI for high-resolution display like Retina, but we do not
need it since we now have our own "Zoom To Fit" work well also for
windows, and the guest is aware of the resolution thanks to UI
information. This change reverts the scaling.

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
block/file-posix: Optimize for macOS

This commit introduces "punch hole" operation and optimizes transfer
block size for macOS.

This commit introduces two additional members,
discard_granularity and opt_io to BlockSizes type in
include/block/block.h. Also, the members of the type are now
optional. Set -1 to discard_granularity and 0 to other members
for the default values.

Thanks to Konstantin Nazarov for detailed analysis of a flaw in an
old version of this change:
https://gist.github.com/akihikodaki/87df4149e7ca87f18dc56807ec5a1bc5#gistcomment-3654667

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
virtio-gpu: Respect UI refresh rate for EDID

This change adds a new member, refresh_rate to QemuUIInfo in
include/ui/console.h. It represents the refresh rate of the
physical display backend, and it is more appropriate than
GUI update interval as the refresh rate which the emulated device
reports:
- sdl may set GUI update interval shorter than the refresh rate
  of the physical display to respond to user-generated events.
- sdl and vnc aggressively changes GUI update interval, but
  a guests is typically not designed to respond to frequent
  refresh rate changes, or frequent "display mode" changes in
  general. The frequency of refresh rate changes of the physical
  display backend matches better to the guest's expectation.

QemuUIInfo also has other members representing "display mode",
which makes it suitable for refresh rate representation. It has
a throttling of update notifications, and prevents frequent changes
of the display mode.

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
ui/cocoa: Take refresh rate into account

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
Use virgl_renderer_borrow_texture_for_scanout

virgl_renderer_borrow_texture_for_scanout is a replacement of
virgl_renderer_resource_get_info and performs configurations such as
swizzle and sRGB decode.

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
ui/cocoa: Update cursor

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
Support Virgil 3D renderer on macOS

Virgil 3D renderer will be available with ui/cocoa on macOS.

The OpenGL backend is NSOpenGLContext for core profile, and
EGL (in particular, ANGLE) for ES profile.

As a notable change, CONFIG_OPENGL_DMABUF is replaced with the
combination of CONFIG_EGL and CONFIG_GBM. In fact, the old
detection of DMA-BUF in configure script was not working
because EGL_MESA_image_dma_buf_export is always defined in
<epoxy/egl.h>. Therefore, the new combination should
behave in the same way.

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
ui/cocoa: Isolate fragments with preprocessor

akihikodaki committed:
edid: Make refresh rate configurable

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki committed:
ui/cocoa: Define types early

akihikodaki committed:
ui/cocoa: Divide ui/cocoa.m

akihikodaki committed:
ui/cocoa: Do not raise keys before QEMU resigns active

ui/cocoa used to raise all keys before it resigns active to prevent a
stuck key problem caused by key up events it does not see while it is
inactive. The problem is solved by checking -[NSEvent modifierFlags] in
commit 6d73bb643aa725348aabe6a885ac5fb0b7f70252, which is better
because it handles the case that key *down* events are missed while it
is inactive.

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

codyd51 committed:
net/macos: implement vmnet-based netdev

This patch implements a new netdev device, reachable via -netdev
vmnet-macos, that’s backed by macOS’s vmnet framework.

The vmnet framework provides native bridging support, and its usage in
this patch is intended as a replacement for attempts to use a tap device
via the tuntaposx kernel extension. Notably, the tap/tuntaposx approach
never would have worked in the first place, as QEMU interacts with the
tap device via poll(), and macOS does not support polling device files.

vmnet requires either a special entitlement, granted via a provisioning
profile, or root access. Otherwise attempts to create the virtual
interface will fail with a “generic error” status code. QEMU may not
currently be signed with an entitlement granted in a provisioning
profile, as this would necessitate pre-signed binary build distribution,
rather than source-code distribution. As such, using this netdev
currently requires that qemu be run with root access. I’ve opened a
feedback report with Apple to allow the use of the relevant entitlement
with this use case:
https://openradar.appspot.com/radar?id=5007417364447232

vmnet offers three operating modes, all of which are supported by this
patch via the “mode=host|shared|bridge” option:

* "Host" mode: Allows the vmnet interface to communicate with other
* vmnet
interfaces that are in host mode and also with the native host.
* "Shared" mode: Allows traffic originating from the vmnet interface to
reach the Internet through a NAT. The vmnet interface can also
communicate with the native host.
* "Bridged" mode: Bridges the vmnet interface with a physical network
interface.

Each of these modes also provide some extra configuration that’s
supported by this patch:

* "Bridged" mode: The user may specify the physical interface to bridge
with. Defaults to en0.
* "Host" mode / "Shared" mode: The user may specify the DHCP range and
subnet. Allocated by vmnet if not provided.

vmnet also offers some extra configuration options that are not
supported by this patch:

* Enable isolation from other VMs using vmnet
* Port forwarding rules
* Enabling TCP segmentation offload
* Only applicable in "shared" mode: specifying the NAT IPv6 prefix
* Only available in "host" mode: specifying the IP address for the VM
within an isolated network

Note that this patch requires macOS 10.15 as a minimum, as this is when
bridging support was implemented in vmnet.framework.

Rebased to commit 3f8d1885e48e4d72eab0688f604de62e0aea7a38 by Akihiko
Odaki.

Signed-off-by: Phillip Tennen <phillip@axleos.com>
Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>
Message-Id: <20210315103209.20870-1-akihiko.odaki@gmail.com>

kov committed:
ui/cocoa: capture all keys and combos when mouse is grabbed

Applications such as Gnome may use Alt-Tab and Super-Tab for different
purposes, some use Ctrl-arrows so we want to allow qemu to handle
everything when it captures the mouse/keyboard.

However, Mac OS handles some combos like Command-Tab and Ctrl-arrows
at an earlier part of the event handling chain, not letting qemu see it.

We add a global Event Tap that allows qemu to see all events when the
mouse is grabbed. Note that this requires additional permissions.

See:

https://developer.apple.com/documentation/coregraphics/1454426-cgeventtapcreate?language=objc#discussion
https://support.apple.com/en-in/guide/mac-help/mh32356/mac

Signed-off-by: Gustavo Noronha Silva <gustavo@noronha.eti.br>

kov committed:
ui/cocoa: add option to swap Option and Command, enable by default

On Mac OS X the Option key maps to Alt and Command to Super/Meta. This change
swaps them around so that Alt is the key closer to the space bar and Meta/Super
is between Control and Alt, like on non-Mac keyboards.

It is a cocoa display option, enabled by default.

Signed-off-by: Gustavo Noronha Silva <gustavo@noronha.eti.br>

akihikodaki commented:
Or can you rebase your changes to `master` of the upstream (and submit to the mailing list)? I'm not really going to submit 'Divide ui/cocoa' patch to the upstream anytime soon and I'm trying to avoid to add patches on top of it.

akihikodaki commented on ui/cocoa/main.m:
```diff
@@ -680,6 +680,13 @@ static void cocoa_display_init(DisplayState *ds, DisplayOptions *opts)
     screen.cursor_show = opts->has_show_cursor && opts->show_cursor;
+    screen.full_grab = opts->u.cocoa.full_grab;
+
+    if (opts->u.cocoa.has_swap_option_command) {
+        screen.swap_option_command = opts->u.cocoa.swap_option_command;
+    } else {
+        screen.swap_option_command = true;
+    }
```
`screen.swap_option_command` always becoms true.

kov replied:
It should not, and I'm pretty sure I tested that both worked. If it has_swap_option_command, then it uses the value that was provided. Otherwise it sets a default.

akihikodaki replied:
I read `has_swap_option_command` as `swap_option_command`. My bad.

akihikodaki commented on ui/cocoa/view.m:
```diff
@@ -207,6 +221,32 @@ - (id)initWithFrame:(NSRect)frameRect
+
+    CGEventMask mask = CGEventMaskBit(kCGEventKeyDown) | CGEventMaskBit(kCGEventKeyUp) | CGEventMaskBit(kCGEventFlagsChanged);
+    eventsTap = CGEventTapCreate(kCGHIDEventTap, kCGHeadInsertEventTap, kCGEventTapOptionDefault,
+                                 mask, (CGEventTapCallBack) handleTapEvent, self);
```
I do not think the type casting of `handleTapEvent` is necessary. Please avoid type casting whenever possible, or the type system will fail to catch a real mistake.

kov replied:
Very used to doing it on Gnome codebases and the Cocoa examples also had it, but I think you are right.

kov replied:
Done.

akihikodaki replied:
The fix is reverted.

kov replied:
Sorry, I probably restored to a previous version of the commits when undoing the squash mistake.

akihikodaki commented on ui/cocoa/view.m:
```diff
@@ -180,6 +180,20 @@ static CGRect compute_cursor_clip_rect(int screen_height,
 
+static CGEventRef handleTapEvent(CGEventTapProxy proxy, CGEventType type, CGEventRef cgEvent, void *userInfo)
+{
+
```
Empty lines are not present at the beginning of other functions. Please remove it.

kov replied:
Definitely a mistake, probably after removing a debug print statement. Will do.

kov commented:
Thanks for your review! If you are not planning on submitting the Divide patch soon I think I'll go ahead and rebase the changes onto master (as soon as I can build qemu again) and submit. Shouldn't cause too many conflicts for your own patch later on I think.

akihikodaki commented on include/ui/cocoa.h:
```diff
@@ -87,6 +90,8 @@ typedef struct {
  */
 - (BOOL) isMouseGrabbed;
 - (BOOL) isAbsoluteEnabled;
+- (BOOL) isFullGrabEnabled;
+- (BOOL) isSwapOptionCommandEnabled;
```
These methods are purely internal ones so remove this from this common header file.

kov replied:
isFullGrabEnabled is actually called from a static function (the callback that the global tap calls), so I think it needs to be here? I removed the other, will push.

akihikodaki replied:
A method is callable from a static function if the implementation appears before the caller, just as a function is in plain old C.

kov commented:
I had somehow squashed both commits together while at the same time not making the changes. Don't git before coffee I guess. This last one should be all set. I had to move the static function around because since it then started complaining that the function was being used without being declared.

kov committed:
ui/cocoa: handle mouse grab properly on window being key changes

Being key is similar to being focused on other platforms. When the
window loses focus at the moment it will still keep the mouse grabbed
unless it is using an absolute pointing device like a USB tablet.

That may happen in several ways as Mac OS does not let Qemu handle some
of the events even when it has a grab. Command-Tab will switch applications
if full-grab is off, and even if it is on you can still switch away from
by using 3-finger gestures on the trackpad to change spaces or open
Mission Control, for instance.

That behaviour is not ideal, as even though Qemu still believes it is
getting all the events, Mac OS has effectively released the mouse.
And it is worse with full-grab turned on, as it will still capture
the key events, making you unable to use the keyboard on the Mac OS
environment.

This change makes Qemu properly release the grab when Mac OS effectively
does it. In addition, it properly automatically regrabs the mouse when
the window becomes key and is fullscreen. While it makes sense to wait
for a click on a non-fullscreen window, as the user may just be moving
the window around, for instance, when it's fullscreen there is no point.

Signed-off-by: Gustavo Noronha Silva <gustavo@noronha.eti.br>

kov commented:
I added one more commit on top of this branch which makes switching between the VM and the host more seamless. Since Mac OS X won't let us handle the 3-finger gestures, we should accept our fates properly and release the grab when Mac OS X effectively does so.
Without this change, if I switch away from the VM using Mission Control but forget to release the grab with C-A-g mouse events will go to the Mac apps but keyboard will still be grabbed due to full-grab. With it, isMouseGrabbed properly turns to false, so full-grab knows it should not be capturing the events.

kov commented:
By the way, I forgot to mention: another way to make it more seamless is to use an absolute pointing device like usb-tablet, but that makes using Gnome worse, as the hot corner becomes very hard to trigger, so I prefer to stick to the virtio-mouse-pci which your run script uses.

akihikodaki reviewed (changes requested):
Some changes for "ui/cocoa: capture all keys and combos when mouse is grabbed" are reverted so I unresolved the corresponding threads. Anything else, including the new commit, looks good to me.

akihikodaki committed:
Merge tag 'patchew/20210120224444.71840-1-agraf@csgraf.de'

kov commented:
Sorry again, thanks for the attention to detail! This one should be correct. Btw I noticed the vmnet commit came along this time, did you remove it from your branch? I removed it from the PR.

akihikodaki committed:
meson: Add implicit_include_directories parameter

This fixes build failures with Xcode 12.5, which was released on April 26,
2021. The Xcode release uses a C++ standard library header file named
"version", for the standard library implementation. The standard library
header file conflicts with the file named "VERSION" at the root of the
source directory and makes the build of the "common" library fail.

By explicitly specifying implicit_include_directories to false for
the library, meson will no longer add the root of the source directory
to the search path of angle bracket inclusions (#include <version>).

The header files in the directory is still referencible because it is
specified for quote inclusions (#include "qemu-options.h"). Such a
conflict avoidance was the intention of the commit introducing quote
inclusion path specifications (commit 9edc19c9
"make: switch from -I to -iquote"). The commit also suggests that it
would also help to "catch any future uses of #include <> for internal
headers."

Signed-off-by: Akihiko Odaki <akihiko.odaki@gmail.com>

akihikodaki commented:
@kov I have just rebased my tree so please cherry-pick your commits to its current head.

kov commented:
Cool, will do so tonight so I can do a bit of testing.

kov commented:
Managed to sneak it in during a meeting ;D

akihikodaki reviewed (approved):
I think these changes are ready for upstreaming. Please rebase to `master` of the upstream.

kov commented:
I have already sent a couple of them upstream. Would you also accept them into your branch?

akihikodaki commented:
@kov I said it is ready for upstreaming, but I found a problem when merging. I replied to your email so please check it out.

akihikodaki commented:
`macos` branch is now rebased to https://patchew.org/QEMU/20210511185538.29344-1-gustavo@noronha.eti.br/ with cherry-picked commit e906a80147b1dc6d4f31b6a08064ef9871a2b76c.
